Images by tag

blakeherrington opened the pull request:
Closes #9

blakeherrington commented:
@bboe This is a quick first pass. I didn't want to do too much since the requirements on this issue are pretty simple. Do we want to go ahead and implement a way for a user to clear out their filters from the UI? Or is navigating home sufficient at this point?

Let me know of your feedback! Thank you!

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,20 @@
 class ImagesController < ApplicationController
   def index
+    _filter_tag = params[:tag]
```
Why did you prefix this variable with an underscore?

blakeherrington replied:
Oopsies. Old habit from the RentLinx days. It has been removed. Addressed here: b7cb9fe

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,20 @@
+      @images = Image.tagged_with([_filter_tag]).order(id: :desc)
+
+      if @images.length > 0
+        return
```
Did you mean to `return` here without rendering anything?

rmacklin commented on config/routes.rb:
```diff
@@ -5,6 +5,7 @@
   # You can have the root of your site routed with "root"
   root 'images#index'
   resources :images, only: [:create, :new, :index, :destroy]
+  get 'images/:tag' => 'images#index'
```
This route would normally be used for an Images#show page. Since this is being used as a filter of the Images#index page, I think it makes more sense to just have the tag be a query parameter for the index route (e.g. `/images?tag=blah`) and remove this line. Your controller wouldn't need to change, you'd just change `image_path(tag)` above to `images_path(tag: tag)`.

blakeherrington replied:
Makes sense. Addressed in a7ce9d3

rmacklin commented:
> Do we want to go ahead and implement a way for a user to clear out their filters from the UI? Or is navigating home sufficient at this point?

Navigating home is sufficient. However, you could still add something to the UI indicating that the index page is being filtered by a tag.

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,20 @@
+    _filter_tag = params[:tag]
+
+    if _filter_tag
+      @images = Image.tagged_with([_filter_tag]).order(id: :desc)
```
You shouldn't need to wrap this in an array before passing it to `tagged_with`

blakeherrington replied:
Addressed in 7bcca0e. Also changed sorting to use created_at since they were in the same file.

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,20 @@
+    end
+
     @images = Image.order(id: :desc)
+
```
I think we can simplify this action a bit so that we don't have to redirect to the same action, we don't need any explicit `return`s, and we don't need to duplicate the `order` parameters. Consider:
```ruby
tag = params[:tag]

if ActsAsTaggableOn::Tag.exists?(name: tag)
  @images = Image.tagged_with(tag)
else
  flash.now[:danger] = "No images found with \"#{tag}\" tag." if tag.present?
  @images = Image.all
end

@images = @images.order(created_at: :desc)
```

For `ActsAsTaggableOn::Tag.exists?(name: tag)` to work properly, you'd want to add:
```ruby
ActsAsTaggableOn.remove_unused_tags = true
```
in config/application.rb.

blakeherrington commented:
Ready for a second look! Changed a couple small behaviors and added some tests as well.

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -1,6 +1,21 @@
+      <%= link_to 'Add New Image', new_image_path, class: 'btn btn-primary', role: 'button' %>
+    </p>
+  </div>
+  <% if ActsAsTaggableOn::Tag.exists?(name: params[:tag]) %>
```
It'd be better not to access `params` from the view. You could make the `tag` variable in the controller action be an instance variable and access it here and on line 12 via `@tag`.

blakeherrington replied:
Makes sense - I can see why that is a better way to go. Addressed in 06834d5

rmacklin commented on app/assets/stylesheets/_filter.scss:
```diff
@@ -0,0 +1,16 @@
+.filter {
+  margin: 8px 0 15px 0;
```
Minor, this is equivalent to the 3-value shorthand: `margin: 8px 0 15px;`

blakeherrington replied:
I was close! Addressed in 2dffe8f

rmacklin commented on app/assets/stylesheets/_filter.scss:
```diff
@@ -0,0 +1,16 @@
+.filter__close:hover {
+  text-decoration: none;
+}
+
```
Minor: extra newline

blakeherrington replied:
Addressed in 639c4a9

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,16 @@
+    end
+
+    @images = @images.order(created_at: :desc)
+
```
Minor: extra newline.

blakeherrington replied:
Grr... I have an issue with pronto on my machine where it won't run if I have any deleted files in my working space. If you have a few minutes sometime soon, it could be helpful to look at together.

Addressed in 210e4d6

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -31,6 +31,27 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '.card', count: 2
+    assert_select '.filter__tag', count: 1, text: 'test1'
+
+    get :index, tag: 'test2'
```
It's not valid to hit multiple actions in a single controller test case.

blakeherrington replied:
Addressed in bdf77dcA

rmacklin commented:
Are you planning to add a flow test for these changes?

blakeherrington commented:
Flow tests have been added in: 6012539

Ready for another look!

bboe commented on app/views/images/index.html.erb:
```diff
@@ -1,6 +1,21 @@
+      <%= link_to 'Add New Image', new_image_path, class: 'btn btn-primary', role: 'button' %>
+    </p>
+  </div>
+  <% if ActsAsTaggableOn::Tag.exists?(name: @tag) %>
```
This query appears to be executing twice. While most databases have a query cache, it would be good to verify that the same query isn't being issued twice. How could we put my concern about duplicate queries at ease?

blakeherrington replied:
I wanted to be checking for sure that the tag name being sent in is a valid tag name (i.e. it is being used on an existing image).

However, I agree this query doesn't need to be executed twice. I have extracted the result into an instance variable which can be reused.

Addressed in c459eb5

bboe commented on app/views/images/index.html.erb:
```diff
@@ -1,6 +1,21 @@
+      <div class="filter">
+        Filtering By:
+          <span class="filter__tag">
+            <span><%= params[:tag] %></span>
```
Looks like there is inconsistent use between `@tag` and `params[:tag]`.

blakeherrington replied:
Oops! Thank you. Addressed in 62a3fa5

bboe commented on app/controllers/images_controller.rb:
```diff
@@ -1,6 +1,17 @@
+      if @tag.present?
+        flash.now[:danger] = "No images found with \"#{@tag}\" tag."
+      end
+      @images = Image.all
```
What happens if Image.all is empty?

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -31,6 +31,41 @@ class ImagesControllerTest < ActionController::TestCase
+                  tag_list: %w(test1 test2 test3))
+    Image.create!(url: 'http://test.com/test.png', tag_list: %w(test1))
+
+    get :index, tag: 'test1'
```
What is the response status?

bboe commented on test/flow_tests/images_flow_test.rb:
```diff
@@ -25,6 +25,31 @@ class ImagesCrudTest < FlowTestCase
+    assert_selector('.card', count: 2)
+  end
+
+  test 'image filtering by invalid tag' do
```
Unless @rmacklin said otherwise, I don't think this test adds anything over the existing controller test. I see the flow tests as the actual flow that someone might go through on your website. Thus the flow tests needn't replicate what is already tested by controllers.

In my mind a simple requirement would be that there are at least 2 HTTP requests made to make a flow test necessary.

blakeherrington replied:
Makes sense. I have removed the unnecessary flow test in fd5fdac

bboe commented:
The PR is looking great. Don't forget that every controller test should do three things:

1) verify the http response status is as expected
2) verify that the http response body contains something expected
3) verify that side-effects either occur or do not for POST/PUT/DELETE requests (GET requests should not have side-effects so there's no need to test for them)

You can find more information here: https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/pairing-information#TOC-Testing-Info

blakeherrington commented:
Ready for a (hopefully) final check.

Added checks to controller tests for response statuses in: 949d586

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -27,10 +28,49 @@ class ImagesControllerTest < ActionController::TestCase
+    get :index, tag: 'test1'
+
+    assert_response :success
+    assert_select '.card', count: 2
```
Am I correct that this is the 2 match case and the other is the 1 match case?

blakeherrington replied:
You are correct. Fixing that!

bboe commented:
Looking really good. I have just one comment regarding test names that might be mixed up. Otherwise great work.

The only other feedback is you have a done of commits while developing. I think at this point you can squash them all down into one or two atomic commits.

blakeherrington commented:
@bboe I have fixed that most recent comment and squashed all my commits down into 1 (I think one made the most sense, I couldn't find a clean line to separate it into 2).

I'll close and merge to master after you give me the 👍

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -27,10 +28,49 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_select '.filter__tag', count: 1, text: 'test1'
+  end
+
+  test 'filter by tag (1 matches)' do
```
I'm being nit-picky. The number changed but not the pluralization of the noun. Feel free to merge as-is though.

bboe commented:
Looks good unless you want to fix a pluralization issue. Great work! 👍

blakeherrington commented:
Yikes! Overlooked that. Haha thanks. 👍
